src/uct/tcp/tcp_ep.c

```diff
@@ -418,6 +418,49 @@ static inline void uct_tcp_ep_ctx_move(uct_tcp_ep_ctx_t *to_ctx,
     memset(from_ctx, 0, sizeof(*from_ctx));
 }
 
+static ucs_status_t uct_tcp_ep_setup_keepalive(uct_tcp_ep_t *ep)
+{
+#if UCT_TCP_EP_KEEPALIVE
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    const int optval       = 1;
+    int idle_sec;
+    int intvl_sec;
+    ucs_status_t status;
+
+    if (!uct_tcp_iface_is_keepalive_enabled(iface)) {
+        return UCS_OK;
+    }
+
+    idle_sec  = ucs_max(1, (int)ucs_time_to_sec(iface->config.keepalive.idle));
+    intvl_sec = ucs_max(1, (int)ucs_time_to_sec(iface->config.keepalive.intvl));
+
+    status = ucs_socket_setopt(ep->fd, SOL_SOCKET, SO_KEEPALIVE,
+                               &optval, sizeof(optval));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucs_socket_setopt(ep->fd, IPPROTO_TCP, TCP_KEEPCNT,
+                               &iface->config.keepalive.cnt,
+                               sizeof(iface->config.keepalive.cnt));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucs_socket_setopt(ep->fd, IPPROTO_TCP, TCP_KEEPIDLE,
+                               &idle_sec, sizeof(idle_sec));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return ucs_socket_setopt(ep->fd, IPPROTO_TCP, TCP_KEEPINTVL,
```

Comment:
set SO_KEEPALIVE last so it will be enabled if all others were ok

Reply:
done